share transport in kubectl

atiratree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
/sig cli

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

When using credentials plugin in a kubeconfig, transport is not reused when creating multiple connections. This applies to many kubectl commands like create/apply/delete/describe when operating on multiple resources. This can cause exhaustion of opened sockets as seen in the following issues. Also please see https://github.com/kubernetes/kubectl/issues/1128#issuecomment-1047744892 and https://github.com/kubernetes/kubectl/issues/1128#issuecomment-1047751466 for more details.

as discussed in https://github.com/kubernetes/kubectl/issues/1128 I am trying to achieve a similar thing in kubectl this PR did https://github.com/kubernetes/kubernetes/pull/105490

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes 
- https://github.com/kubernetes/kubectl/issues/1128 
- https://github.com/kubernetes/kubectl/issues/1152

#### Special notes for your reviewer:

Please see https://github.com/kubernetes/kubernetes/pull/108459#issuecomment-1058525691

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->

Relevant PRs:
- https://github.com/kubernetes/kubernetes/pull/97811
- https://github.com/kubernetes/kubernetes/pull/97855

atiratree committed:
share http client to reuse transport in ConfigFlags

- add ToHTTPClient function to RESTClientGetter interface

atiratree committed:
share http client to reuse transport in Builder

atiratree committed:
share http client in Factory

atiratree committed:
share http client in describe

atiratree committed:
use Factory http Client in exec and attach

atiratree committed:
adjust TestFactory to pass fakeClient to Factory and refactor methods

accordingly

atiratree commented:
still need to figure out a later change of config (stdin detection for exec credentials provider) after the transport has been established

aojea commented on staging/src/k8s.io/kubectl/pkg/cmd/attach/attach.go:
```diff
@@ -128,7 +130,7 @@ type RemoteAttach interface {
 func DefaultAttachFunc(o *AttachOptions, containerToAttach *corev1.Container, raw bool, sizeQueue remotecommand.TerminalSizeQueue) func() error {
 	return func() error {
-		restClient, err := restclient.RESTClientFor(o.Config)
+		restClient, err := restclient.RESTClientForConfigAndClient(o.Config, o.HttpClient)
```
don't you have the same effect just moving the constructor before the closure?

```go
// DefaultAttachFunc is the default AttachFunc used
func DefaultAttachFunc(o *AttachOptions, containerToAttach *corev1.Container, raw bool, sizeQueue remotecommand.TerminalSizeQueue) func() error {
restClient, err := restclient.RESTClientFor(o.Config)
		if err != nil {
			return err
		}
	return func() error {
				req := restClient.Post().
			Resource("pods").
			Name(o.Pod.Name).
			Namespace(o.Pod.Namespace).
			SubResource("attach")
```

atiratree replied:
Unfortunately not, since the config and client is not yet initialized

atiratree committed:
Allow passing stdinInUse back to ConfigFlags (RESTClientGetter)

to correctly setup Exec Credentials Provider

This callback has to be added to support the following scenario

1. ConfigFlags and Factory is created
2. DynamicClient, DiscoveryClient / OpenAPIGetter is obtained from
   Factory and used. http.Client gets cached in Factory
3. Builder is created and usage of stdin is detected via
   - implicitely by FilenameParam use in many commands
   - detected explicitely by replace.go, set_env.go and Builder is
     notified
4. Builder has knowledge of all stdins and can fire StdinMultiUseError
5. cached http.Client needs to be reset by calling onStdinInUse on
   ConfigFlags (RESTClientGetter) to support interactive mode of Exec
   Credentials Provider and prevent stdin collisions
6. New clients can be obtained from the Factory and Builder with correctly
   configured and cached http.Client

atiratree commented:
To correctly implement Exec Credentials Provider which behaviour was handled by Builder so far I had to implement a callback in `Allow passing stdinInUse back to ConfigFlags` commit.

It is  a bit unfortunate since it makes the code more complex but I wrote an explanation there.  Alternative would be to move any Client invocation before the builder which would require a big code reshuffling across a lot of kubectl commands. I think this is a better alternative.

Please let me know any feedback on this. Commit message:

```
    Allow passing stdinInUse back to ConfigFlags (RESTClientGetter)
    
    to correctly setup Exec Credentials Provider
    
    This callback has to be added to support the following scenario
    
    1. ConfigFlags and Factory is created
    2. DynamicClient, DiscoveryClient / OpenAPIGetter is obtained from
       Factory and used. http.Client gets cached in Factory
    3. Builder is created and usage of stdin is detected via
       - implicitely by FilenameParam use in many commands
       - detected explicitely by replace.go, set_env.go and Builder is
         notified
    4. Builder has knowledge of all stdins and can fire StdinMultiUseError
    5. cached http.Client needs to be reset by calling onStdinInUse on
       ConfigFlags (RESTClientGetter) to support interactive mode of Exec
       Credentials Provider and prevent stdin collisions
    6. New clients can be obtained from the Factory and Builder with correctly
       configured and cached http.Client

```

atiratree commented on staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go:
```diff
@@ -75,6 +75,8 @@ type RESTClientGetter interface {
 	// ToHTTPClient returns HTTPClient for sharing transport
 	ToHTTPClient() (*http.Client, error)
+	// SetStdinInUse sets a stdinInUse flag to true and resets clients
+	SetStdinInUse()
```
Alternatively we could pass a bool here to be more future proof. Thoughts?

soltysh replied:
Personally, I'm not a fan of this approach, can we split that bit out of this PR and handle just the transport share in this PR?

atiratree replied:
- even though this approach is not that pretty, it is a bit hard to come up with an alternative (https://github.com/kubernetes/kubernetes/pull/108459#discussion_r826386790). Maybe We could rename the interface or create additional interface on top of RESTClientGetter?
- spliting the PR is not that easy since we would introduce a regression and could block stdin and not advertise it back to http client

liggitt replied:
> Personally, I'm not a fan of this approach, can we split that bit out of this PR and handle just the transport share in this PR?

+1, I don't think this bit need to escape out of resource.Builder

brianpursley commented on staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go:
```diff
@@ -123,6 +128,12 @@ type ConfigFlags struct {
+	httpClientLock sync.Mutex
+
+	// Affects behaviour of Exec Credentials Provider.
+	stdinInUse bool
```
It seems like there should be a way not to have to add `stdinInUse` to the config.

What if you were to do one of these things instead:

1. Keep track of the configuration that was used when httpClient was created and if the config is different then don't reuse the existing httpClient, but create a new one instead

2. Keep a map of httpClients to cache config-specific instances of httpClient (similar to how `tlsTransportCache` works).

3. Keep two httpClient instances, one with stdin available and one without

It just seems like this PR would be a lot simpler if you can leave all the `stdinInUse` stuff inside builder.go where it is currently.

atiratree replied:
Unfortunately there are couple of problems with these approaches. The main issue is that we are using the clients before we have parsed / recieved all the inputs for the kubectl command (please take a look at https://github.com/kubernetes/kubernetes/pull/108459#issuecomment-1058525691 for more details). And kubectl decides later if we will be using stdin or not.

1. As mentioned, the problem is that we amend the config manually later so a source of this information is needed for such tracking (like `stdinInUse`).
2. same as 1. we need to update/callback the Factory manually that we wish for a config change. It is not received externally. Also as discussed here there might be some problems with caching: https://github.com/kubernetes/kubernetes/pull/108274#discussion_r812046486
3. This could be an option but not sure how much feasible. Problem is that the caller would have to decide if it needs a client with or without stdin enabled. The caller would have to track and use this properly. I think there is bigger chance of an error and larger code maintanace with this approach.

I understand the current approach is a bit harder to grasp, but it is not such a big change (basically a callback). I have tried to document this in the code, but I welcome any additional feedback on how to improve this.

brianpursley commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -238,8 +243,8 @@ func describerMap(clientConfig *rest.Config) (map[schema.GroupKind]ResourceDescr
 // Kubernetes types.
-func DescriberFor(kind schema.GroupKind, clientConfig *rest.Config) (ResourceDescriber, bool) {
-	describers, err := describerMap(clientConfig)
+func DescriberFor(kind schema.GroupKind, clientConfig *rest.Config, httpClient *http.Client) (ResourceDescriber, bool) {
```
Similar to a previous comment I made, I'm not sure how important it is to maintain the signatures of public functions, or if we want to still support `DescribeFor` that creates its own HTTP client and add a new `DescribeForConfigAndClient` that accepts a client.

I'm not saying it definitely needs to be a separate function, but just asking the question

atiratree replied:
I am also not sure how much the backwards compatibility is important but I have added methods with `AndClient` suffix to be sure. Also applied this to `GenericDescriberFor` as well.

brianpursley commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -183,8 +188,8 @@ func (npw *nestedPrefixWriter) WriteLine(a ...interface{}) {
-func describerMap(clientConfig *rest.Config) (map[schema.GroupKind]ResourceDescriber, error) {
-	c, err := clientset.NewForConfig(clientConfig)
+func describerMap(clientConfig *rest.Config, client *http.Client) (map[schema.GroupKind]ResourceDescriber, error) {
+	c, err := clientset.NewForConfigAndClient(clientConfig, client)
```
What if client is nil?

Maybe something like this?

```
if client == nil {
    c, err = clientset.NewForConfig(clientConfig)
} else {
    c, err = clientset.NewForConfigAndClient(clientConfig, client)
}
```

If not, then at a minimum it should probably return an error in `DescriberFor` if httpClient is nil.

atiratree replied:
changed to this construct for safety

brianpursley reviewed (commented):
Left a few comments.

I wonder if it can be simplified further.  It seems like leaving stdinInUse inside builder would go a long way towards keeping it simpler.

atiratree commented on staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go:
```diff
@@ -284,7 +356,7 @@ func (f *ConfigFlags) toDiscoveryClient() (discovery.CachedDiscoveryInterface, e
 	discoveryCacheDir := computeDiscoverCacheDir(filepath.Join(cacheDir, "discovery"), config.Host)
 
-	return diskcached.NewCachedDiscoveryClientForConfig(config, discoveryCacheDir, httpCacheDir, time.Duration(10*time.Minute))
+	return diskcached.NewCachedDiscoveryClientForConfig(config, httpClient, discoveryCacheDir, httpCacheDir, time.Duration(10*time.Minute))
```
I found on a closer examination, it is not very easy to share the transport with the cached discovery as it is wrapping the the transport for caching. At least not without changing a visibility of some classes across packages. So I have decided to omit the discovery client for now. It should not be a big issue as it is just one additional socket (with cache) and we try to reuse the discovery client in kubectl.

atiratree commented:
/test pull-kubernetes-integration

atiratree committed:
fix print test

atiratree commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

atiratree committed:
add ConfigFlags test for stdinInUse and clients initialization

atiratree committed:
add Kubectl Factory integration tests for sharing transport

atiratree commented:
I have added unit and integration tests. 

I am not sure if this a good candidate for e2e tests. Please let me know if we should also run them and detect opened sockets?

soltysh commented:
/hold
Spoke with Filip on slack, this requires more discussion and we don't want to rush such a big refactor hours before the freeze.

atiratree commented:
/test pull-kubernetes-integration

dovholuknf commented:
I'd be interested in learning how this ends up getting implemented. Being able to provide the transport is a key feature. A while back we tied into the DialFunc from the restConfig. Being able to provide a single transport that gets used/reused would be great

atiratree commented:
/test pull-kubernetes-e2e-kind

atiratree commented:
/remove-lifecycle stale

liggitt commented:
This got pointed to as a possible fix for an issue that required us to revert removal of a couple in-tree credential plugins - https://github.com/kubernetes/kubernetes/issues/111911

I would really like to see this PR made as safe and minimal as possible so we can merge it early in 1.26 and backport it to 1.23/1.24/1.25.

liggitt commented on staging/src/k8s.io/cli-runtime/pkg/resource/builder.go:
```diff
@@ -215,6 +227,8 @@ func NewBuilder(restClientGetter RESTClientGetter) *Builder {
 	return newBuilder(
 		restClientGetter.ToRESTConfig,
```
Would it be more contained to 1) replicate the httpClient caching logic inside resource.Builder, 2) make it opt-in, and 3) only enable httpClient reuse from contexts we know are short-lived/safe (like kubectl `factoryImpl#NewBuilder()`)?

That would avoid leaking `stdinInUse` ugliness outside `resource.Builder`, ensure we don't regress `resource.Builder` behavior for callers other than kubectl, and make this more likely to be able to be backported

liggitt commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -239,7 +250,13 @@ func describerMap(clientConfig *rest.Config) (map[schema.GroupKind]ResourceDescr
 // Kubernetes types.
 func DescriberFor(kind schema.GroupKind, clientConfig *rest.Config) (ResourceDescriber, bool) {
-	describers, err := describerMap(clientConfig)
+	return DescriberForClient(kind, clientConfig, nil)
```
any callers to this call site will still spawn new transports, right? do we have any callers to this site and can we eliminate them?

liggitt commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -183,8 +188,14 @@ func (npw *nestedPrefixWriter) WriteLine(a ...interface{}) {
+	var c *clientset.Clientset
+	var err error
+	if client == nil {
+		c, err = clientset.NewForConfig(clientConfig)
```
any callers that hit this code path will still spawn new transports, right? do we know if any kubectl callers hit this code path?

liggitt commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -252,14 +269,31 @@ func DescriberFor(kind schema.GroupKind, clientConfig *rest.Config) (ResourceDes
 // GenericDescriberFor returns a generic describer for the specified mapping
 // that uses only information available from runtime.Unstructured
 func GenericDescriberFor(mapping *meta.RESTMapping, clientConfig *rest.Config) (ResourceDescriber, bool) {
+	return GenericDescriberForClient(mapping, clientConfig, nil)
```
same question about whether any callers invoke this

liggitt commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -252,14 +269,31 @@ func DescriberFor(kind schema.GroupKind, clientConfig *rest.Config) (ResourceDes
+	var err error
 	// used to fetch the resource
-	dynamicClient, err := dynamic.NewForConfig(clientConfig)
+	if httpClient == nil {
```
if httpClient is nil, we should at least construct it once in this function and use the NewForConfigAndClient paths instead of constructing two transports here and below for the events client

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/factory_client_access.go:
```diff
@@ -100,46 +117,58 @@ func (f *factoryImpl) RESTClient() (*restclient.RESTClient, error) {
 
 func (f *factoryImpl) ClientForMapping(mapping *meta.RESTMapping) (resource.RESTClient, error) {
-	cfg, err := f.clientGetter.ToRESTConfig()
+	clientConfig, err := f.ToRESTConfig()
```
nit: if possible, keep the variable names the same to minimize diff for backports

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/util/factory_client_access.go:
```diff
@@ -100,46 +117,58 @@ func (f *factoryImpl) RESTClient() (*restclient.RESTClient, error) {
 
 func (f *factoryImpl) UnstructuredClientForMapping(mapping *meta.RESTMapping) (resource.RESTClient, error) {
-	cfg, err := f.clientGetter.ToRESTConfig()
+	clientConfig, err := f.ToRESTConfig()
```
same comment about keeping existing variable names for backports

liggitt commented on test/integration/kubectl/factory_share_transport_test.go:
```diff
@@ -0,0 +1,306 @@
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestKubectlFactoryShareTransport(t *testing.T) {
```
might also be worth exercising the clients created by the resource.Builder

that's what was tripping up repeated construction from `apply` (reproducer integration test as part of https://github.com/kubernetes/kubernetes/issues/111911 at https://github.com/liggitt/kubernetes/commits/exec-auth-reproducer demonstrating one transport, and therefore a new connection, for each item in an applied file when using exec auth)

dims commented:
@atiratree i don't see any responses to the last questions from Jordan. needs a rebase as well!

dims commented:
If you still need this PR then please rebase, if not, please close the PR

aojea commented:
@atiratree @liggitt I know you are busy, but just wanted to get your thoughts about this, it seems we were close and is a bit sad to not be able to get this in

dovholuknf commented:
It'd be nice if all the functionality uses the same transport too. Right now exec doesn't appear to use the same mechanism, there's probably other functionality that doesn't use it either.  Ideally anything that needs to talk to the kubernetes API would all use the same http transport or better yet the same `net.Dialer`

atiratree commented:
I am sorry that this has slipped.

I will see what I can do, and try to pick this up in the coming weeks.

atiratree commented:
There is almost no chance this will make it to 1.27. Will try again to pick this up for 1.28

atiratree commented:
/remove-lifecycle stale
